_posts/2023-09-19-virtual-thread-1.adoc

```diff
@@ -0,0 +1,281 @@
+
+[source, java]
+----
+@Path("/greetings")
+public class VirtualThreadApp {
+
+  @RestClient RemoteService service;
+
+  @GET
+  @RunOnVirtualThread
+  public String process() {
+    // Runs on a virtual thread because the
+    // method uses the @RunOnVirtualThread annotation.
+
+    // `service` is a rest client, it executes an I/O operation
+    var response = service.greetings(); // Blocking, but this time, it 
+                                        // does neither block the carrier thread
+                                        // nor the OS thread.
+                                        // Only the virtual thread is blocked.
+	return response.toUpperCase();
+  }
+
+}
+----
+
+It's the code from the first snippet (the imperative one), but its execution model is closer to the reactive one.
+This snippet works as described because the Quarkus REST client is virtual-thread-firendly.
+
+## Five things you need to know before using virtual threads for everything
+
+Well, you probably see this coming.
+There is no free lunch.
+You need to know a few things before utilizing virtual threads for everything.
+These are the reasons why, currently, there is no global switch to run exclusively on virtual threads in Quarkus.
+
+### 1. Pinning 
+
+As described above, when a virtual thread executes a blocking operation, it gets unmounted from the carrier thread, preventing the carrier thread from being blocked.
+However, sometimes, the virtual thread cannot be unmounted because its state cannot be stored in the heap. 
+It happens when the thread holds a monitor lock or has a native call in the stack. 
+In this case, the carrier thread is blocked, so the OS thread is blocked. 
+
+Unfortunately, as of today, lots of Java libraries are pinning the carrier thread. 
+The Quarkus team and Red Hat, in general, have patched many libraries (such as Narayana (the transaction manager of Quarkus) or Hibernate ORM) to avoid pinning. 
+However, when you use a library, be careful. 
+It will take time until all the code gets reworked in a more virtual-thread-friendly way.
+
+### 2. Monopolisation
+
+As for the reactive model, if the virtual thread executes intensive and long computation, it monopolizes that carrier.
+The virtual thread scheduler is not preemptive.
+So it cannot interrupt a running thread. 
+It needs to wait for an I/O or the completion of the computation.
+Until then, this carrier thread cannot execute other virtual threads.
+Using a dedicated platform thread pool might be wiser when executing long computations. 
+
+### 3. Carrier thread pool elasticity
+
+When there is pinning or monopolization, the JVM may create new carrier threads to avoid having too many unscheduled virtual threads. 
+
+These creations are creating platform/OS threads. 
+So, it's expensive and uses memory.
+You especially need to pay attention to the second point. You may hit the memory limit if you run on low resources and your code is not very virtual-thread-friendly. 
```

Comment:
... virtual-thread-friendly, meaning that you should always check for pinning, monopolization, and memory usage. If you don't, in a container with memory constraints, the application can be killed.

Reply:
Fixed.